File: docs/tasks/task_format.ts
/**
 * TypeScript interfaces mirroring docs/tasks/task_format.py
 *
 * Source of truth: docs/tasks/task_format.py
 * This file provides a TypeScript representation of the same schema for
 * use in TypeScript codebases or documentation tooling.
 */

// Task/Feature status codes
export type Status = 
  | "+" // Done
  | "~" // In Progress
  | "-" // Pending
  | "?" // Blocked
  | "="; // Deferred

export interface Feature {
  id: string;
  status: Status;
  title: string;
  description: string;
  plan: string;
  context: string[];
  acceptance: string[];
  dependencies?: string[];
  rejection?: string;
}

export interface Task {
  id: number;
  status: Status;
  title: string;
  description: string;
  features: Feature[];
  rejection?: string;
}

export interface ProjectRequirement {
  id: number;
  status: Status;
  description: string;
  tasks: number[];
}

export interface ProjectSpec {
  id: string;
  title: string;
  description: string;
  path: string;
  repo_url: string;
  requirements: ProjectRequirement[];
}

// Utility type guards (optional helpers)
export function isStatus(val: unknown): val is Status {
  return val === "+" || val === "~" || val === "-" || val === "?" || val === "=";
}

File: docs/tasks/task_format.py
from typing import Dict, TypedDict, List, Literal

try:
    from typing import NotRequired
except ImportError:
    from typing_extensions import NotRequired

Status = Literal[
    "+", # Done
    "~", # Crunching
    "-", # Pending
    "?", # Blocked
    "=" # Deferred
]

class Feature(TypedDict):
    id: str
    status: Status
    title: str
    description: str
    plan: str
    context: List[str]
    acceptance: List[str]
    blockers: NotRequired[List[str]] # ["{story_id}.{feature_id}","{story_id}"]
    rejection: NotRequired[str]

class Story(TypedDict):
    id: str
    status: Status
    title: str
    description: str
    features: List[Feature]
    blockers: NotRequired[List[str]] # ["{story_id}.{feature_id}","{story_id}"]
    rejection: NotRequired[str]
    featureIdToDisplayIndex: Dict[str,int]

class ProjectRequirement(TypedDict):
    id: int
    status: Status
    description: str
    stories: List[str]

class ProjectSpec(TypedDict):
    id: str
    title: str
    description: str
    path: str
    repo_url: str
    requirements: List[ProjectRequirement]
    storyIdToDisplayIndex: Dict[str,int]
